File: Homework/Hw12/hw12.java
import java.util.*;

public class hw12 {
    private int[] parent;

    public hw12(int n) {
        parent = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            parent[i] = i;
        }
    }

    private int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private boolean union(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if (rootX != rootY) {
            parent[rootY] = rootX;
            return true;
        }
        return false;
    }

    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        List<int[]> edges = new ArrayList<>();

        // Adding virtual node 0 connections with wells cost
        for (int i = 0; i < n; i++) {
            edges.add(new int[] {wells[i], 0, i + 1});
        }

        // Adding pipes
        for (int[] pipe : pipes) {
            edges.add(new int[] {pipe[2], pipe[0], pipe[1]});
        }

        // Sort edges based on cost
        Collections.sort(edges, Comparator.comparingInt(a -> a[0]));

        int totalCost = 0;
        int edgesUsed = 0;

        // Process each edge and form the MST
        for (int[] edge : edges) {
            if (union(edge[1], edge[2])) {
                totalCost += edge[0];
                edgesUsed++;
                if (edgesUsed == n) {
                    break;
                }
            }
        }

        return totalCost;
    }

    public static void main(String[] args) {
        int n = 2;
        int[] wells = {1, 1};
        int[][] pipes = {{1, 2, 1}, {1, 2, 2}};
        hw12 solution = new hw12(n);
        int result = solution.minCostToSupplyWater(n, wells, pipes);
        System.out.println("Minimum cost to supply water: " + result);
    }
}

File: Homework/Hw12/hw12.py
from typing import List

class Solution:
    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:
        parent = list(range(n + 1))
        
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(x, y):
            rootX = find(x)
            rootY = find(y)
            if rootX != rootY:
                parent[rootY] = rootX
                return True
            return False
        
        # Create edges including virtual node 0 to each house with the cost of building a well
        edges = []
        for i in range(n):
            edges.append((wells[i], 0, i + 1))
        
        # Add existing pipes
        for house1, house2, cost in pipes:
            edges.append((cost, house1, house2))
        
        # Sort edges based on the cost
        edges.sort()
        
        total_cost = 0
        edges_used = 0
        
        # Process each edge in increasing order of cost
        for cost, house1, house2 in edges:
            if union(house1, house2):
                total_cost += cost
                edges_used += 1
                if edges_used == n:
                    break
        
        return total_cost

# Example usage:
if __name__ == "__main__":
    sol = Solution()
    n = 2
    wells = [1, 1]
    pipes = [[1, 2, 1], [1, 2, 2]]
    result = sol.minCostToSupplyWater(n, wells, pipes)
    print("Minimum cost to supply water:", result)
